### PR TITLE
Add support for RestGetMapping cancellation

### PR DESCRIPTION
**Pull request by fcofdez**

### CODE REVIEW

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Comment by fcofdez** [comment]
@elasticmachine update branch

**Review by DaveCTurner — COMMENTED** [review]
Would it be possible also to check for cancellation on each iteration of the loop in `Metadata#findMappings`? Iterating the mappings is pretty expensive, it'd be great if we could abort that loop early where applicable.

**Comment by fcofdez** [comment]
> Would it be possible also to check for cancellation on each iteration of the loop in Metadata#findMappings? Iterating the mappings is pretty expensive, it'd be great if we could abort that loop early where applicable.

I've added support for a more fine grained task cancellation checks, but I run into some issues while testing it. 
I added a fake `MapperPlugin` that allowed injecting a custom `getFieldFilter` that's used during the get mapping transport action. The fake `getFieldFilter` blocks the thread until some conditions are meet, the problem comes when the http connection shares the same event-loop thread that the blocked transport action. In that scenario the closed connection event is not executed until the action is unblocked, meaning that the task is not cancelled. I've decided to move the `TransportGetMappingsAction` to the `GENERIC` thread pool, but I'm not sure if that change could cause problems elsewhere. Let me know what you think @DaveCTurner. Maybe there's a better way to test it.

**Comment by DaveCTurner** [comment]
I think it would be acceptable just to extend `MetadataTests#testFindMappings` to verify that `Metadata#findMappings` checks for cancellation the right number of times. I suppose that wouldn't cover us against passing the wrong thing (e.g. a no-op check) into `findMappings` but I think we can take that risk.

Why is the check a `Consumer<String>` rather than just a `Runnable`? I didn't see anything using the index name, although I didn't check very thoroughly.

**Comment by fcofdez** [comment]
> I think it would be acceptable just to extend MetadataTests#testFindMappings to verify that Metadata#findMappings checks for cancellation the right number of times. I suppose that wouldn't cover us against passing the wrong thing (e.g. a no-op check) into findMappings but I think we can take that risk.

Thanks for the feedback, that seems like a good idea 👍 

> Why is the check a Consumer<String> rather than just a Runnable? I didn't see anything using the index name, although I didn't check very thoroughly.

That was just in case it could be useful in the future for other cases, but I'll use a `Runnable` instead.

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java [hunk 1]
```diff
@@ -374,6 +375,21 @@ public boolean equalsAliases(Metadata other) {
      */
     public ImmutableOpenMap<String, MappingMetadata> findMappings(String[] concreteIndices,
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java:376]: I think we should inline this overload, there's only a couple of other usages of this method in production code and it's a pretty small thing for the callers to pass in a no-op check if that's what they want.

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java [hunk 2]
```diff
@@ -73,14 +75,15 @@ public RestChannelConsumer prepareRequest(final RestRequest request, final NodeC
...
-
-            @Override
-            protected void processResponse(GetMappingsResponse getMappingsResponse) {
-                final long startTimeMs = threadPool.relativeTimeInMillis();
-                // Process serialization on MANAGEMENT pool since the serialization of the raw mappings to XContent can be too slow to
-                // execute on an IO thread
-                threadPool.executor(ThreadPool.Names.MANAGEMENT).execute(
+        final HttpChannel httpChannel = request.getHttpChannel();
+        return channel -> new RestCancellableNodeClient(client, httpChannel).admin().indices().getMappings(getMappingsRequest,
+            new RestActionListener<>(channel) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java:80]: This is quite similar to `DispatchingRestToXContentListener`, with deficiencies on both sides:

- here we have a timeout but `DispatchingRestToXContentListener` doesn't
- `DispatchingRestToXContentListener` checks for cancellation after dispatch but here we don't
- Here we don't have a proper `ToXContentObject`, we call `startObject()` and `endObject()` ourselves, but `DispatchingRestToXContentListener` requires a `ToXContentObject`.

What do you think about moving to `DispatchingRestToXContentListener` here? I think it'd just need a little adapter object to implement the timeout and the outer `XContent` object wrapper, see e.g. `RestClusterStateResponse`.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a couple more style suggestions.

**Comment by fcofdez** [comment]
@elasticmachine update branch

File: server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java (hunk 2 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/rest/action/admin/indices/RestGetMappingAction.java]: Addressed in https://github.com/elastic/elasticsearch/pull/72234/commits/e2d96baf2c381300d66cbc415623f23dc93f6bc1

File: server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java (hunk 1 above)
Comment (fcofdez) [reply server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java]: Addressed in https://github.com/elastic/elasticsearch/pull/72234/commits/e2d96baf2c381300d66cbc415623f23dc93f6bc1

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**